src/itemStruct.ts

```diff
@@ -0,0 +1,111 @@
+import * as vscode from "vscode"
+import { getAutocompleteDump } from "./autocompleteDump"
+import { getApiDump, UNCREATABLE_TAGS } from "./dump"
+
+export class ItemStructCompletionProvider implements vscode.CompletionItemProvider {
+    itemStructNames: Promise<vscode.CompletionItem[]>
+    itemStructs: Promise<{ [name: string]: vscode.CompletionItem[] }>
+
+    constructor() {
+        this.itemStructs = (async () => {
+            const autocompleteDump = await getAutocompleteDump()
+            const apiDump = await getApiDump()
+            const itemStructs: { [name: string]: vscode.CompletionItem[] } = {}
+            for (const itemStruct of autocompleteDump.ItemStruct) {
+                itemStructs[itemStruct.name] = [
+                    ...itemStruct.properties.filter(property => property.static).map(property => {
+                        const item = new vscode.CompletionItem(property.name, vscode.CompletionItemKind.Field)
+                        item.detail = `(property) ${itemStruct.name}.${property.name}: ${property.type}`
+                        item.documentation = new vscode.MarkdownString(`${property.description ? property.description + "\n\n" : ""}[Developer Reference](https://developer.roblox.com/en-us/api-reference/datatype/${itemStruct.name})`)
+                        return item
+                    }),
+                    ...itemStruct.functions.filter(func => func.static).map(func => {
+                        const insertText = new vscode.SnippetString(`${func.name}(`)
+
+                        const params = []
+                        for (const paramIndex in func.parameters) {
+                            if (func.parameters[paramIndex]) {
+                                const param = func.parameters[paramIndex]
+                                const paramText = `${param.name}${param.optional ? "?" : ""}: ${param.type || "unknown"}`
+                                params.push(paramText)
+
+                                // Create a snippet if the parameters are definable (eg. Instance.new())
+                                let paramInsertText = null
+                                if (param.constraint) {
+                                    const constraintSplit = param.constraint.split(":")
+                                    const objectType = constraintSplit[0]
+                                    const constraint = constraintSplit[1] || "any"
+
+                                    paramInsertText = apiDump.Classes.filter(klass => {
+                                        if (objectType === "Instance") {
+                                            if (constraint === "any") {
+                                                return true
+                                            } else if (constraint === "isScriptCreatable") {
+                                                const tags = klass.Tags
+                                                if (tags) {
```

Comment:
Explicit `!== undefined` check here as well as wherever else you might use it.

Reply:
Apologies, I seem to keep forgetting!